### ISSUE TITLE
Ambiguitiy in causal-mask in scaled_dot_product_attention

### ISSUE THREAD

**Issue by bigeagle** [pytorch/pytorch#103082]
### 🐛 Describe the bug

When `is_causal=True`, `torch.nn.functional.scaled_dot_product_attention` is defined as an efficient implementation of the following

```
# Q: (N, ..., L, E)
# K: (N, ..., S, E)
# V: (N, ..., S, Ev)
attn_mask = torch.ones((L, S), dtype=torch.bool).tril(diagonal=0)
attn_mask = attn_mask.maksed_fill(not attn_mask, -float('inf'))
attn_weight = torch.softmax((Q @ K.transpose(-2, -1) / math.sqrt(Q.size(-1))) + attn_mask, dim=-1)
out = attn_weight @ V
```

If `L != S`, the definition of attn_mask may be ambiguous. 

For example, during the auto-regressive generation process of a GPT-like model, when kv-cache is enabled, usually `L=1`. In this case, 

```python
In [1]: attn_mask = torch.ones((1, S), dtype=torch.bool).tril(diagonal=0); attn_mask
Out[1]: tensor([[ True, False, False, False, False, False, False, False]])
``` 

However, the correct `attn_mask` in this scenario should be `[True, True, ..., True]`.

I'm not sure it's a bug or feature. I guess there must be a reason why the causal mask is designed in the current way. It would be helpful to discuss.

### Versions

2.0

**Comment by drisspg** [comment]
Im going to close this as a we do have a mechanism for specifying which alignment to use: https://github.com/pytorch/pytorch/blob/0246429866104294197ca3a49049e3bf9172de9f/torch/nn/attention/bias.py#L33
